### ISSUE TITLE
Compiler enforced annotations for non total functions.

### ISSUE THREAD

**Issue by Josef-Vonasek** [luna/luna#292]
### General Summary

Every function that is non total (non exhaustive match, function that can panic (rust terminology)), should be annotated or you get compiler error.

### Example

Lets say the annotation symbol would be `!` - then:

```elm
get : Vector a -> Int -> a
get ix = if ix >= self.size then panic "Index out of bounds" else ...  
```
will throw a compiler error 
```
Function `get` can panic for certain inputs, please annotate it with `!`.

Hint: rename `get` to `get!`
```

### Motivation

Non total functions are the number one source of bugs, yet sometimes necessary evil, when time constrains and convenience are at play. 

There are three types of software

1. security critical  - NASA, medical instruments, crypto currencies
2. common software - web, other services, most of today software
3. data science - (and scripting)

Luna definitely wants to target 3rd group where non total functions are not such an issue. Yet it would be a mistake to close the doors for group 1, where one simple bug can cost millions of dollars!

I think that an explicit non-totality annotation is a great compromise to allow developers from group 1 and 3 enjoy the same ecosystem:

 * It is concise: `vector[1][2]` vs `vector[1]![2]!`
 * The safe behavior is the default one: `head : Vector a -> Maybe a`
 * It is explicit: it is easy to spot the source of potential bug and easy to search with `ctrl+f`
 * The related compiler error messages are easy to understand.

Note: I think that few obvious functions like division could circumvent this rule, even though I personally would not mind writing `/!`

**Comment by wdanilo** [comment]
As I've written on our chat, this is a very interesting idea. It's true that our **first** target group are data scientists, but we definitely don't want to sacrifice language security for it. There are few points from my side here:

1. It is not always possible to check if a function is total. Sure we can check for exhaustive pattern matches and as time passes add some more heuristics but we will never get the real solution here (please correct me if I'm wrong). If we go this way and add new heuristics with time and require function names to have `!`, wich each of such heuristics we will break existing code and it's bad.

2. When thinking about this enhancement proposal I started to consider another way to do it as well - just automatically marking functions that we've discovered to be non-total with some kind of a warning icon in both text as well as visual representation. The problem with gluing `!` to names is that if you fix a function and it is no longer non-total, you sometimes have to refactor **a lot** of functions that used it. What's even more important, if you fix such function in a library, a lot of other libraries and user projects could stop to compile due to changed name. 

To sum this up - the totality checking is an important feature. We cannot discover all possible situations and we should consider it some important metric of how good the code is. If you write a code it's important to improve its robustness and security over time. Nobody is perfect, so bugs will always appear and libraries and people will make non-total functions on purpose, often due to lack of time. Improving the security should not always lead IMO to a breaking API change, so requiring function name change is questionable to me. Nevertheless, it's an important proposal and I'm sure we should indicate such situations clearly to users.

**Comment by Josef-Vonasek** [comment]
1. Yes, I am talking specifically about:
  * Non exhaustive pattern matching 
  * Using a other function which is nontotal (`panic`) 
  * Calling code from other language (FFI). 
  You still won't be able to detect infinite loops (`while foo ... -- oops, foo is always true`), but that's not even purpose of this proposal, because it is not generally possible in Turing complete language.

2. For the reason above, it should be possible to annotate any function with `!` even if the compiler thinks that the function is well defined for all inputs, and therefore if you for example remove a call to some c++ code from from a function `foo!`, you can now export both `foo!` and `foo` and won't complain.

**Comment by wdanilo** [comment]
Ok, it makes sense. Now I need to better understand why requirement to explicitly mark such functions is better than just having an warning icon next to it in text and on nodes. It gives the same information and it’s easier for users, right?

**Comment by Josef-Vonasek** [comment]
Except that most of the time you want to export two versions of the same function: one safe and one unsafe.

This means you will like end up with something like `head` and `maybeHead` and we all know that people will favor the unsafe version, just because it is shorter and promoted by auto completion.

**Comment by iamrecursion** [comment]
Alternatively you could be more explicit about it, having `head` and `unsafeHead`. This brings the same benefits of annotation, but is far more explicit. Personally, I think `!` would add to syntactic noise in what we intend to be a fairly clean language. 

In any case, we would have to decide on exactly what the totality checker looks for. Idris has some [interesting work in this regard[(http://docs.idris-lang.org/en/latest/tutorial/theorems.html#totality-checking), allowing detection of even potentially non-well-founded recursion (with programmer overrides).

**Comment by Josef-Vonasek** [comment]
> Personally, I think ! would add to syntactic noise in what we intend to be a fairly clean language.

It is definitely less verbose than `unsafe` ;)

**Comment by iamrecursion** [comment]
Verbosity isn't the same thing as syntactic noise!

**Comment by iamrecursion** [comment]
If there is interest in further pursuing this kind of thing, I recommending writing up an RFC and submitting it to the [luna-rfcs](https://github.com/luna/luna-rfcs) repo.
